src/components/timeline/TrackPowerGraph.js

```diff
@@ -129,34 +130,74 @@ class TrackPowerCanvas extends React.PureComponent<CanvasProps> {
       ctx.fillStyle = '#73737388'; // Grey 50 with transparency.
       ctx.beginPath();
 
-      // The x and y are used after the loop.
-      let x = 0;
-      let y = 0;
-      let firstX = 0;
-      for (let i = sampleStart; i < sampleEnd; i++) {
-        // Create a path for the top of the chart. This is the line that will have
-        // a stroke applied to it.
-        x = (samples.time[i] - rangeStart) * millisecondWidth;
+      const getX = (i) =>
+        Math.round((samples.time[i] - rangeStart) * millisecondWidth);
+      const getPower = (i) => {
         const sampleTimeDeltaInMs =
           i === 0 ? interval : samples.time[i] - samples.time[i - 1];
-        const unitGraphCount =
-          samples.count[i] / sampleTimeDeltaInMs / countRangePerMs;
+        return samples.count[i] / sampleTimeDeltaInMs;
+      };
+      const getY = (rawY) => {
+        const unitGraphCount = rawY / countRangePerMs;
         // Add on half the stroke's line width so that it won't be cut off the edge
         // of the graph.
-        y =
+        return Math.round(
           innerDeviceHeight -
-          innerDeviceHeight * unitGraphCount +
-          deviceLineHalfWidth;
-        if (i === 0) {
-          // This is the first iteration, only move the line, do not draw it. Also
-          // remember this first X, as the bottom of the graph will need to connect
-          // back up to it.
-          firstX = x;
-          ctx.moveTo(x, y);
-        } else {
+            innerDeviceHeight * unitGraphCount +
+            deviceLineHalfWidth
+        );
+      };
+
+      // The x and y are used after the loop.
+      const firstX = getX(sampleStart);
+      let x = firstX;
+      let y = getY(getPower(sampleStart));
+
+      // For the first sample, only move the line, do not draw it. Also
+      // remember this first X, as the bottom of the graph will need to connect
+      // back up to it.
+      ctx.moveTo(x, y);
+
+      // Create a path for the top of the chart. This is the line that will have
+      // a stroke applied to it.
+      for (let i = sampleStart + 1; i < sampleEnd; i++) {
+        const powerValues = [getPower(i)];
+        x = getX(i);
+        y = getY(powerValues[0]);
+        ctx.lineTo(x, y);
+
+        // If we have multiple samples to draw on the same horizontal pixel, we
+        // process all of them together to save time:
+        // - We draw the first and last samples to ensure the display is
+        //   correct if there are sampling gaps.
+        // - For the values in between, we only draw the min and max values,
+        //   to draw a vertical line covering all the other sample values.
```

Comment:
it would be also good to mention the "min/max decimation" name for future reference and also maybe we can put the link that Julien mentioned in his comment? First I was a bit confused with the algorithm but looking at the description made it more obvious (I was confused with how many points we draw for each pixel)

Reply:
I'm adding "with max-min decimation algorithm" to the wording of the comment. I was hoping "we only draw the min and max values, to draw a vertical line covering all the other sample values." in my comment was descriptive enough of how the algorithm worked, but I guess not. Thanks for suggesting the comment improvement.

Reply:
Thanks for inproving the comment!